File: Server/src/plugin/dialogue/LumbridgeCookDialogue.java
package plugin.dialogue;

import org.crandor.game.content.dialogue.DialoguePlugin;
import org.crandor.game.content.dialogue.FacialExpression;
import org.crandor.game.node.entity.npc.NPC;
import org.crandor.game.node.entity.player.Player;
import org.crandor.plugin.InitializablePlugin;
import org.crandor.game.node.item.Item;

/**
 * Represents the dialogue plugin used for the lumbridge cook.
 * @author 'Vexia
 * @version 1.0
 */
@InitializablePlugin
public final class LumbridgeCookDialogue extends DialoguePlugin {

	/**
	 * Represents the milk item.
	 */
	private static final Item MILK = new Item(1927, 1);

	/**
	 * Represents the flour item.
	 */
	private static final Item FLOUR = new Item(1933, 1);

	/**
	 * Represents the eggs item.
	 */
	private static final Item EGG = new Item(1944, 1);

	/**
	 * Constructs a new {@code LumbridgeCookDialogue} {@code Object}.
	 */
	public LumbridgeCookDialogue() {
		/**
		 * empty.
		 */
	}

	/**
	 * Constructs a new {@code LumbridgeCookDialogue} {@code Object}.
	 * @param player the player.
	 */
	public LumbridgeCookDialogue(Player player) {
		super(player);
	}

	@Override
	public DialoguePlugin newInstance(Player player) {
		return new LumbridgeCookDialogue(player);
	}

	@Override
	public boolean open(Object... args) {
		npc = (NPC) args[0];
		if (player.getQuestRepository().getQuest("Cook's Assistant").getStage(player) == 10) {
			if (player.getSavedData().getQuestData().getCookAssist("milk") && player.getSavedData().getQuestData().getCookAssist("flour") && player.getSavedData().getQuestData().getCookAssist("egg")) {
				interpreter.sendDialogues(npc, FacialExpression.HAPPY, "You've brought me everything I need! I am saved!", "Thank you!");
				stage = 956;
				return true;
			}
			interpreter.sendDialogues(npc, FacialExpression.SAD, "How are you getting on with finding the ingredients?");
			stage = 1000;
			return true;
		} if (player.getQuestRepository().getQuest("Cook's Assistant").getStage(player) == 0) {
			interpreter.sendDialogues(npc, FacialExpression.SAD, "What am I to do?");
			stage = 1;
			return true;
		} if (player.getQuestRepository().getQuest("Cook's Assistant").getStage(player) == 100) {
			interpreter.sendDialogues(npc, FacialExpression.HAPPY,"Hello friend, how is the adventuring going?");
			stage = 0;
		}
		stage = 0;
		return true;
	}

	@Override
	public boolean handle(int interfaceId, int buttonId) {
		switch (stage) {
			case 0: //Dialogue for after Cook's Assistant
				interpreter.sendOptions("Select an Option", "I am getting strong and mighty.", "I keep on dying.", "Can I use your range?");
				stage = 5000;
				break;
			case 1:
				if (player.getQuestRepository().getQuest("Cook's Assistant").getStage(player) == 0) {
					interpreter.sendOptions("Select an Option", "What's wrong?", "Can you make me a cake?", "You don't look very happy.", "Nice hat!");
					stage++;
					break;
				}
			case 2:
				switch (buttonId) {
					case 1:
						interpreter.sendDialogues(player, FacialExpression.NEUTRAL, "What's wrong?");
						stage = 10;
						break;
					case 2:
						interpreter.sendDialogues(player, FacialExpression.ASKING, "You're a cook, why don't you bake me a cake?");
						stage = 20;
						break;
					case 3:
						interpreter.sendDialogues(player, FacialExpression.NEUTRAL, "You don't look very happy.");
						stage = 30;
						break;
					case 4:
						interpreter.sendDialogues(player, FacialExpression.HAPPY, "Nice hat!");
						stage = 40;
						break;

				}
				break;

			//Dialogues for 'Can you bake me a cake?'
			case 20:
				interpreter.sendDialogues(npc, FacialExpression.SAD, "*sniff* Don't talk to me about cakes...");
				stage++;
				break;
			case 21:
				interpreter.sendDialogues(player, FacialExpression.NEUTRAL, "What's wrong?");
				stage = 10;
				break;

			//Dialogue for 'You don't look very happy'
			case 30:
				interpreter.sendDialogues(npc, FacialExpression.SAD, "No, I'm not. The world is caving in around me - I am", "overcome by dark feelings of impending doom.");
				stage = 21;
				break;


			//Dialogue for 'Nice Hat!'
			case 40:
				interpreter.sendDialogues(npc, FacialExpression.SAD, "Er, thank you. It's a pretty ordinary cook's hat, really.");
				stage++;
				break;
			case 41:
				interpreter.sendDialogues(player, FacialExpression.HAPPY, "Still, it suits you. The trousers are pretty special too.");
				stage++;
				break;
			case 42:
				interpreter.sendDialogues(npc, FacialExpression.SAD, "It's all standard-issue cook's uniform.");
				stage++;
				break;
			case 43:
				interpreter.sendDialogues(player, FacialExpression.HAPPY, "The whole hat, apron, stripy trousers ensemble. It", "works. It makes you looks like a real cook.");
				stage++;
				break;
			case 44:
				interpreter.sendDialogues(npc, FacialExpression.ANGRY, "I AM a real cook! I haven't got time to be chatting", "about culinary fashion. I'm in desperate need of help!");
				stage = 21;
				break;

			//Main Dialogue
			case 10:
				interpreter.sendDialogues(npc, FacialExpression.SCARED, "Oh dear, oh dear, oh dear, I'm in a terrible terrible", "mess! It's the Duke's birthday today, and I should be", "making him a lovely big birthday cake.");
				stage++;
				break;
			case 11:
				interpreter.sendDialogues(npc, FacialExpression.SAD, "I've forgotten to buy the ingredients. I'll never get", "them in time now. He'll sack me! What will I do? I have", "four children and a goat to look after. Would you help", "me? Please?");
				stage++;
				break;
			case 12:
				interpreter.sendOptions("Select an Option", "I'm always happy to help a cook in distress.", "I can't right now, Maybe later.");
				stage++;
				break;
			case 13:
				switch (buttonId) {
					case 1:
						interpreter.sendDialogues(player, FacialExpression.HAPPY, "Yes, I'll help you.");
						stage = 100;
						break;
					case 2:
						interpreter.sendDialogues(player, FacialExpression.NEUTRAL, "No, I don't feel like it. Maybe later.");
						stage = 14;
						break;

				}
				break;

			//If the Player does not want to help the Chef
			case 14:
				interpreter.sendDialogues(npc, FacialExpression.SAD, "Fine. I always knew you Adventurer types were callous", "beasts. Go on your merry way!");
				stage++;
				break;
			case 15:
				end();
				break;

			//If the Player does help the Chef
			case 100:
				player.getQuestRepository().getQuest("Cook's Assistant").start(player);
				interpreter.sendDialogues(npc, FacialExpression.HAPPY, "Oh thank you, thank you. I need milk, an egg and", "flour. I'd be very grateful if you can get them for me.");
				stage++;
				break;
			case 101:
				interpreter.sendDialogues(player, FacialExpression.NEUTRAL, "So where do I find these ingredients then?");
				stage++;
				break;

			//Where do I get the Ingredients?
			case 102:
				interpreter.sendOptions("Select an Option", "Where do I find some flour?", "How about milk?", "And eggs? Where are they found?", "Actually, I know where to find this stuff.");
				stage++;
				break;
			case 103:
				switch (buttonId) {
					case 1:
						interpreter.sendDialogues(npc, FacialExpression.HAPPY, "There is a Mill fairly close, Go North and then West.", "Mill Lane Mill is just off the road to Draynor. I", "usually get my flour from there.");
						stage = 110;
						break;
					case 2:
						interpreter.sendDialogues(npc, FacialExpression.HAPPY, "There is a cattle field on the other side of the river,", "just across the road from the Groats' Farm.");
						stage = 120;
						break;
					case 3:
						interpreter.sendDialogues(npc, FacialExpression.HAPPY, "I normally get my eggs from the Groats' farm, on the", "other side of the river.");
						stage = 130;
						break;
					case 4:
						interpreter.sendDialogues(player, FacialExpression.HAPPY, "Actually, I know where to find this stuff.");
						stage = 15;
						break;
				}
				break;

			case 110:
				interpreter.sendDialogues(npc, FacialExpression.HAPPY, "Talk to Millie, she'll help, she's a lovely girl and a fine", "Miller..");
				stage = 140;
				break;
			case 120:
				interpreter.sendDialogues(npc, FacialExpression.HAPPY, "Talk to Gillie Groats, she looks after the Dairy cows -", "she'll tell you everything you need to know about", "milking cows!");
				stage = 140;
				break;
			case 130:
				interpreter.sendDialogues(npc, FacialExpression.HAPPY, "But any chicken should lay eggs.");
				stage = 140;
				break;

			//There is alternative dialogue as the last option, this could be condensed, but this is the easiest solution right now.
			case 140:
				interpreter.sendOptions("Select an Option", "Where do I find some flour?", "How about milk?", "And eggs? Where are they found?", "I've got all the information I need. Thanks.");
				stage++;
				break;
			case 141:
				switch (buttonId) {
					case 1:
						interpreter.sendDialogues(npc, FacialExpression.HAPPY, "There is a Mill fairly close, Go North and then West.", "Mill Lane Mill is just off the road to Draynor. I", "usually get my flour from there.");
						stage = 110;
						break;
					case 2:
						interpreter.sendDialogues(npc, FacialExpression.HAPPY, "There is a cattle field on the other side of the river,", "just across the road from the Groats' Farm.");
						stage = 120;
						break;
					case 3:
						interpreter.sendDialogues(npc, FacialExpression.HAPPY, "I normally get my eggs from the Groats' farm, on the", "other side of the river.");
						stage = 130;
						break;
					case 4:
						interpreter.sendDialogues(player, FacialExpression.HAPPY, "I've got all the information I need. Thanks.");
						stage = 15;
						break;
				}
				break;

			//Code for giving the chef the items
			case 1000:
				boolean gave = false;
				if (player.getInventory().contains(1927, 1) && !player.getAttribute("gaveMilk", false)) {
					player.getInventory().remove(MILK);
					interpreter.sendDialogues(player, FacialExpression.HAPPY, "Here's a bucket of milk.");
					player.getSavedData().getQuestData().setCooksAssistant("milk", true);
					player.getSavedData().getQuestData().setCooksAssistant("gave", true);
					player.setAttribute("gaveMilk", true);
					gave = true;
					break;
				}
				if (player.getInventory().contains(1933, 1) && !player.getAttribute("gaveFlour", false)) {
					player.getInventory().remove(FLOUR);
					interpreter.sendDialogues(player, FacialExpression.HAPPY, "Here's a pot of flour.");
					player.getSavedData().getQuestData().setCooksAssistant("flour", true);
					player.getSavedData().getQuestData().setCooksAssistant("gave", true);
					player.setAttribute("gaveFlour", true);
					gave = true;
					break;
				}
				if (player.getInventory().contains(1944, 1) && !player.getAttribute("gaveEgg", false)) {
					player.getInventory().remove(EGG);
					player.getSavedData().getQuestData().setCooksAssistant("egg", true);
					player.getSavedData().getQuestData().setCooksAssistant("gave", true);
					player.setAttribute("gaveEgg", true);
					interpreter.sendDialogues(player, FacialExpression.HAPPY, "Here's a fresh egg.");
					gave = true;
					break;
				}
				if (!gave) {
					if (!player.getSavedData().getQuestData().getCookAssist("gave")) {
						interpreter.sendDialogue("You still need to get:", "A bucket of milk. A pot of flour. An egg.");
						stage = 1100;
					} else {
						interpreter.sendDialogues(npc, FacialExpression.SAD, "Thanks for the ingredients you have got so far please get", "the rest quickly. I'm running out of time! The Duke", "will throw me into the streets!");
						stage = 1201;
					}
					break;
				}
				stage = 1200;
				break;

			//If the player has not found any ingredients or if only some of the ingredients have been given to the chef
			case 1100:
				interpreter.sendOptions("Select an Option", "I'll get right on it.", "Can you remind me how to find these things again?");
				stage++;
				break;
			case 1101:
				switch (buttonId) {
					case 1:
						interpreter.sendDialogues(player, FacialExpression.HAPPY, "I'll get right on it.");
						stage = 15;
						break;
					case 2:
						interpreter.sendDialogues(player, FacialExpression.NEUTRAL, "Can you remind me how to find these things again?.");
						stage = 140;
						break;
				}
				break;
			case 1200:
				interpreter.sendDialogues(npc, FacialExpression.SAD, "Thanks for the ingredients you have got so far please get", "the rest quickly. I'm running out of time! The Duke", "will throw me into the streets!");
				stage++;
				break;
			case 1201:
				String[] messages = new String[4];
				messages[0] = "You still need to get:";
				if (!player.getSavedData().getQuestData().getCookAssist("milk")) {
					messages[1] = "A bucket of milk. ";
				}
				if (!player.getSavedData().getQuestData().getCookAssist("flour")) {
					messages[2] = "A pot of flour. ";
				}
				if (!player.getSavedData().getQuestData().getCookAssist("egg")) {
					messages[3] = "An egg. ";
				}
				StringBuilder builder = new StringBuilder();
				if (messages[1] != null) {
					builder.append(messages[1]);
				}
				if (messages[2] != null) {
					builder.append(messages[2]);
				}
				if (messages[3] != null) {
					builder.append(messages[3]);
				}
				if (builder.length() != 0) {
					interpreter.sendDialogue(messages[0], builder.toString());
					stage = 1100;
				} else {
					interpreter.sendDialogues(npc, FacialExpression.HAPPY, "You've brought me everything I need! I am saved!", "Thank you!");
					stage = 956;
				}
				break;

			//Final Cooks Assistant Dialogue
			case 956:
				interpreter.sendDialogues(player, FacialExpression.HAPPY, "So do I get to go to the Duke's Party?");
				stage++;
				break;
			case 957:
				interpreter.sendDialogues(npc, FacialExpression.SAD, "I'm afraid not, only the big cheeses get to dine with the", "Duke.");
				stage++;
				break;
			case 958:
				interpreter.sendDialogues(player, FacialExpression.NEUTRAL, "Well, maybe one day I'll be important enough to sit on", "the Duke's table.");
				stage++;
				break;
			case 959:
				interpreter.sendDialogues(npc, FacialExpression.NEUTRAL, "Maybe, but I won't be holding my breath.");
				stage++;
				break;

			//End Quest 'Cook's Assistant'
			case 960:
				end();
				player.getQuestRepository().getQuest("Cook's Assistant").finish(player);
				break;

			//Dialogue options after Cook's Assistant
			case 5000:
				switch (buttonId) {
					case 1:
						interpreter.sendDialogues(player, FacialExpression.HAPPY, "I am getting strong and mighty. Grr.");
						stage = 5100;
						break;
					case 2:
						interpreter.sendDialogues(player, FacialExpression.SAD, "I keep on dying.");
						stage = 5200;
						break;
					case 3:
						interpreter.sendDialogues(player, FacialExpression.ASKING, "Can I use your range?");
						stage = 5300;
						break;
				}
				break;

			case 5100:
				interpreter.sendDialogues(npc, FacialExpression.HAPPY, "Glad to hear it!");
				stage = 15;
				break;
			case 5200:
				interpreter.sendDialogues(npc, FacialExpression.HAPPY, "Ah, well, at least you keep coming back to life too!");
				stage = 15;
				break;
			case 5300:
				interpreter.sendDialogues(npc, FacialExpression.HAPPY, "Go ahead! It's a very good range; it's better than most" ,"other ranges.");
				stage++;
				break;
			case 5301:
				interpreter.sendDialogues(npc, FacialExpression.HAPPY, "It's called the Cook-o-Matic 100 and it uses a combination","of state-of-the-art temperature regulation and magic.");
				stage++;
				break;
			case 5302:
				interpreter.sendDialogues(player, FacialExpression.ASKING, "Will it mean my food will burn less often?");
				stage++;
				break;
			case 5303:
				interpreter.sendDialogues(npc, FacialExpression.HAPPY, "As long as the food is fairly easy to cook in the first place!");
				stage = 15;
				break;
		}
		return true;
	}

	@Override
	public int[] getIds() {
		return new int[] { 278 };
	}
}

File: Server/src/main/kotlin/rs09/game/content/quest/free/cooksassistant/LumbridgeCookDialogue.kt
package rs09.game.content.quest.free.cooksassistant

import core.game.content.dialogue.DialoguePlugin
import core.game.content.dialogue.FacialExpression
import core.game.node.entity.npc.NPC
import core.game.node.entity.player.Player
import core.game.node.item.Item
import core.plugin.Initializable

/**
 * Dialogue for the Lumbridge Cook.
 * @author Qweqker
 */

@Initializable
class LumbridgeCookDialogue (player: Player? = null) : DialoguePlugin(player){

    //Item declaration
    val EMPTY_BUCKET = 1925
    val MILK = 1927
    val EMPTY_POT = 1931
    val FLOUR = 1933
    val EGG = 1944

    //Default settings for the Cook checking for ingredients in the player's inventory
    var gave = false
    var leftoverItems = ""


    override fun open(vararg args: Any?): Boolean {
        npc = args[0] as NPC
        if(player?.questRepository?.getQuest("Lost Tribe")?.getStage(player) == 10){
            player("Did you see what happened in the cellar?")
            stage = 0
            return true
        }
        if (player?.questRepository?.getQuest("Cook's Assistant")!!.getStage(player) <= 0) { //If the player has ot started cook's assistant
            npc(FacialExpression.SAD, "What am I to do?")
            stage = 0
            return true
        } else if (player?.questRepository?.getQuest("Cook's Assistant")!!.getStage(player) in 10..99) { //During the Cook's Assistant Quest
            if (player.getAttribute("cooks_assistant:all_submitted", false) || (player.getAttribute("cooks_assistant:milk_submitted", false) && player.getAttribute("cooks_assistant:flour_submitted", false) && player.getAttribute("cooks_assistant:egg_submitted", false))){ //If the player has handed all the ingredients to the chef but did not continue the dialogue
                npc(FacialExpression.HAPPY, "You've brought me everything I need! I am saved!", "Thank you!")
                stage = 200
                return true
            } else { //If the player has not handed all the items to the chef
                npc(FacialExpression.SAD, "How are you getting on with finding the ingredients?")
                gave = false //Resetting to false here to prevent dialogue skipping later
                stage = 100
                return true
            }
        }
        //After completing Cook's Assistant
        npc(FacialExpression.HAPPY, "Hello friend, how is the adventuring going?")
        stage = 300
        return true
    }

    override fun handle(interfaceId: Int, buttonId: Int): Boolean {
        if(player.questRepository.getQuest("Lost Tribe").getStage(player) == 10){
            when(stage){
                //Lost Tribe
                0 -> npc("Last night I was in the kitchen and I heard a noise","from the cellar. I opened the trapdoor and saw a","creature dart into a hole in the wall.").also { stage++ }
                1 -> npc("It looked a bit like a goblin, but it had big bulging eyes.","It wasn't wearing armour, but it had this odd helmet","with a light on it.").also { stage++ }
                2 -> npc("The tunnel was too dark for me to follow it, so I went","to tell the Duke. But when we went down to the cellar","the hole had been blocked up, and no one believes me.").also { stage++ }
                3 -> player("I believe you.").also { stage++ }
                4 -> npc("Thank you, ${player.name}! If you can convince the Duke","I'm telling the truth then we can get to the bottom of","this mystery.").also { stage = 1000; player.questRepository.getQuest("Lost Tribe").setStage(player,20) }
                5 -> end()
            }
            return true
        }
        when (stage) {
            0 -> options("What's wrong?", "Can you make me a cake?", "You don't look very happy.", "Nice hat!").also { stage++ }
            1 -> when(buttonId) {
                1 -> player(FacialExpression.NEUTRAL, "What's wrong?").also { stage = 10 }
                2 -> player(FacialExpression.ASKING, "You're a cook, why don't you bake me a cake?").also { stage = 20 }
                3 -> player(FacialExpression.NEUTRAL,"You don't look very happy.").also { stage = 30; }
                4 -> player(FacialExpression.HAPPY, "Nice hat!").also { stage = 40 }
            }

            //What's wrong?
            10 -> npc(FacialExpression.SCARED, "Oh dear, oh dear, oh dear, I'm in a terrible terrible", "mess! It's the Duke's birthday today, and I should be", "making him a lovely big birthday cake.").also { stage++ }
            11 -> npc(FacialExpression.SAD, "I've forgotten to buy the ingredients. I'll never get", "them in time now. He'll sack me! What will I do? I have", "four children and a goat to look after. Would you help", "me? Please?").also { stage++ }
            12 -> options("I'm always happy to help a cook in distress.", "I can't right now, Maybe later.").also { stage++ }
            13 -> when(buttonId) {
                1 -> player(FacialExpression.HAPPY, "Yes, I'll help you.").also { stage = 50 }
                2 -> player(FacialExpression.ANNOYED, "No, I don't feel like it. Maybe later.").also { stage++ }
            }

            //I can't right now
            14 -> npc(FacialExpression.SAD, "Fine. I always knew you Adventurer types were callous", "beasts. Go on your merry way!").also { stage = 1000 }

            //Can you make me a cake?
            20 -> npc(FacialExpression.SAD, "*sniff* Don't talk to me about cakes...").also { stage++ }
            21 -> player(FacialExpression.NEUTRAL, "What's wrong?").also { stage = 10}

            //You don't look very happy
            30 -> npc(FacialExpression.SAD, "No, I'm not. The world is caving in around me - I am", "overcome by dark feelings of impending doom.").also { stage++ }
            31 -> options("What's wrong?","I'd take the rest of the day off if I were you.").also { stage++ }
            32 -> when(buttonId) {
                1 -> player(FacialExpression.NEUTRAL, "What's wrong?").also { stage = 10}
                2 -> player(FacialExpression.NEUTRAL,"I'd take the rest of the day off if I were you.").also { stage++ }
            }
            33 -> npc(FacialExpression.SAD,"No, that's the worst thing I could do. I'd get in terrible","trouble.").also { stage++ }
            34 -> player(FacialExpression.ASKING,"Well maybe you need to take a holiday...").also { stage++ }
            35 -> npc(FacialExpression.SAD,"That would be nice, but the Duke doesn't allow holidays","for core staff").also { stage++ }
            36 -> player(FacialExpression.SUSPICIOUS,"Hmm, why not run away to the sea and start a new","life as a Pirate?").also { stage++ }
            37 -> npc(FacialExpression.SAD,"My wife gets sea sick, and I have an irrational fear of","eyepatches. I don't see it working myself.").also { stage++ }
            38 -> player(FacialExpression.NEUTRAL,"I'm afraid I've run out of ideas.").also { stage++ }
            39 -> npc(FacialExpression.SAD,"I know I'm doomed.").also { stage = 21 }

            //Nice hat!
            40 -> npc(FacialExpression.SAD, "Er, thank you. It's a pretty ordinary cook's hat, really.").also { stage++ }
            41 -> player(FacialExpression.HAPPY, "Still, it suits you. The trousers are pretty special too.").also { stage++ }
            42 -> npc(FacialExpression.SAD, "It's all standard-issue cook's uniform.").also { stage++ }
            43 -> player(FacialExpression.HAPPY, "The whole hat, apron, stripy trousers ensemble. It", "works. It makes you looks like a real cook.").also { stage++ }
            44 -> npc(FacialExpression.ANGRY, "I AM a real cook! I haven't got time to be chatting", "about culinary fashion. I'm in desperate need of help!").also { stage = 21 }

            //Yes, I'll help you
            50 -> npc(FacialExpression.HAPPY, "Oh thank you, thank you. I need milk, an egg and", "flour. I'd be very grateful if you can get them for me.").also{ player?.questRepository?.getQuest("Cook's Assistant")?.start(player!!); stage++ }
            51 -> player(FacialExpression.NEUTRAL, "So where do I find these ingredients then?").also { stage = 60 }

            //Where do I find these ingredients?
            60 -> options("Where do I find some flour?","How about milk?","And eggs? Where are they found?","Actually, I know where to find this stuff.").also { stage++ }
            61 -> when(buttonId) {
                1 -> npc(FacialExpression.NEUTRAL,"There is a Mill fairly close, go North and then West.","Mill Lane Mill is just off the road to Draynor. I","usually get my flour from there.").also {stage = 70 }
                2 -> npc(FacialExpression.NEUTRAL,"There is a cattle field on the other side of the river,","just across the road from the Groats' Farm.").also { stage = 71 }
                3 -> npc(FacialExpression.NEUTRAL,"I normally get my eggs from the Groats' farm, on the","other side of the river.").also { stage = 73 }
                4 -> player(FacialExpression.NEUTRAL,"Actually, I know where to find this stuff.").also { stage = 1000 }
            }

            //Where do I find some flour?
            70 -> npc(FacialExpression.SUSPICIOUS,"Talk to Millie, she'll help, she's a lovely girl and a fine","Miler. Make sure you take a pot with you for the flour","though, " + if (player.inventory.contains(EMPTY_POT, 1)) "you've got one on you already." else "there should be one on the table in here.").also { stage = 80 }

            //How about milk?
            71 -> npc(FacialExpression.SUSPICIOUS,"Talk to Gillie Groats, she looks after the Dairy cows -","she'll tell you everything you need to know about","milking cows!").also { stage++ }
            72 ->
                if (player.inventory.contains(EMPTY_BUCKET , 1)) {
                        npc(FacialExpression.NEUTRAL,"You'll need an empty bucket for the milk itself. I do see", "you've got a bucket with you already luckily!").also { stage = 80 }
                } else {
                    npc(FacialExpression.NEUTRAL,"You'll need an empty bucket for the milk itself. The", "general store just north of the castle will sell you one", "for a couple of coins.").also { stage = 80 }
                }

            //And Eggs?
            73 -> npc(FacialExpression.NEUTRAL,"But any chicken should lay eggs.").also { stage = 80 }

            //Alternative menu for "Where do I find these ingredients?"
            80 -> options("Where do I find some flour?","How about milk?","And eggs? Where are they found?","I've got all the information I need. Thanks.").also { stage++ }
            81 -> when(buttonId) {
                1 -> npc(FacialExpression.NEUTRAL,"There is a Mill fairly close, go North and then West.","Mill Lane Mill is just off the road to Draynor. I","usually get my flour from there.").also {stage = 70 }
                2 -> npc(FacialExpression.NEUTRAL,"There is a cattle field on the other side of the river,","just across the road from the Groats' Farm.").also { stage = 71 }
                3 -> npc(FacialExpression.NEUTRAL,"I normally get my eggs from the Groats' farm, on the","other side of the river.").also { stage = 73 }
                4 -> player(FacialExpression.NEUTRAL,"I've got all the information I need. Thanks.").also { stage = 1000 }
            }

            100 ->
                if (!player.getAttribute("cooks_assistant:milk_submitted", false) && player.inventory.contains(MILK, 1)) {
                    player.setAttribute("/save:cooks_assistant:milk_submitted", true).also {
                        player(FacialExpression.HAPPY, "Here's a bucket of milk.")
                        player.inventory.remove(Item(MILK))
                        gave = true
                        stage = 100
                    }
                } else if (!player.getAttribute("cooks_assistant:flour_submitted", false) && player.inventory.contains(FLOUR, 1)) {
                    player.setAttribute("/save:cooks_assistant:flour_submitted", true).also {
                        player(FacialExpression.HAPPY, "Here's a pot of flour.")
                        player.inventory.remove(Item(FLOUR))
                        gave = true
                        stage = 100
                    }
                } else if (!player.getAttribute("cooks_assistant:egg_submitted", false) && player.inventory.contains(EGG, 1)) {
                    player.setAttribute("/save:cooks_assistant:egg_submitted", true).also {
                        player(FacialExpression.HAPPY, "Here's a fresh egg.")
                        player.inventory.remove(Item(EGG))
                        gave = true
                        stage = 100
                    }
                } else {
                    if (gave) {

                        //If this is the first time the player gave an item (or items) to the Lumbridge cook
                        if (!player.getAttribute("cooks_assistant:submitted_some_items", false)) {
                            player.setAttribute("/save:cooks_assistant:submitted_some_items", true)
                        }

                        //If the player has now handed in all the ingredients
                        if (player.getAttribute("cooks_assistant:milk_submitted", false) && player.getAttribute("cooks_assistant:flour_submitted", false) && player.getAttribute("cooks_assistant:egg_submitted", false)) {
                            npc(FacialExpression.HAPPY, "You've brought me everything I need! I am saved!", "Thank you!").also { player.setAttribute("/save:cooks_assistant:all_submitted",true); stage = 200 }
                        } else {
                            npc(FacialExpression.WORRIED,"Thanks for the ingredients you have got so far, please get","the rest quickly. I'm running out of time! The Duke","will throw me into the streets!").also { stage = 151 }
                        }

                    } else { //If the player did not give an item to the Lumbridge cook

                        //If the player also has never submitted anything before
                        if (!player.getAttribute("cooks_assistant:submitted_some_items", false)) {
                            player(FacialExpression.NEUTRAL, "I haven't gotten any of them yet, I'm still looking.").also { stage = 155 }
                        } else {
                            options("I'll get right on it.", "Can you remind me how to find these things again?").also { stage = 161 }
                        }
                    }
                }

            //If the player has submitted some ingredients but not all of them
            150 -> npc(FacialExpression.WORRIED,"Thanks for the ingredients you have got so far, please get","the rest quickly. I'm running out of time! The Duke","will throw me into the streets!").also { stage++ }

            //Checking what the player has left over
            151 -> leftoverItems = "".also{
                if (!player.getAttribute("cooks_assistant:milk_submitted", false)){
                    leftoverItems += "A bucket of milk. "
                }
                if (!player.getAttribute("cooks_assistant:flour_submitted", false)){
                    leftoverItems += "A pot of flour. "
                }
                if (!player.getAttribute("cooks_assistant:egg_submitted", false)){
                    leftoverItems += "An egg."
                }
                if (leftoverItems != ""){
                    sendDialogue("You still need to get:",leftoverItems).also {stage = 160}
                } else {
                    npc(FacialExpression.HAPPY, "You've brought me everything I need! I am saved!", "Thank you!").also { player.setAttribute("/save:cooks_assistant:all_submitted",true); stage = 200 }
                }
            }

            //If the player has yet to collect or hand in any of the ingredients
            155 -> npc(FacialExpression.WORRIED,"Please get the ingredients quickly. I'm running out of","time! The Duke will throw me into the streets!").also { stage++ }
            156 -> sendDialogue("You still need to get:", "A bucket of milk. A pot of flour. An egg.").also { stage = 160 }


            // Menu after checking the items handed in
            160 -> options("I'll get right on it.","Can you remind me how to find these things again?").also { stage++ }
            161 -> when(buttonId) {
                1 -> player(FacialExpression.NEUTRAL,"I'll get right on it.").also { stage = 1000 }
                2 -> player(FacialExpression.ASKING,"Can you remind me how to find these things again?").also { stage = 60 }
            }

            //Final Cooks Assistant Dialogue
            200 -> player(FacialExpression.HAPPY, "So do I get to go to the Duke's Party?").also { stage++ }
            201 -> npc(FacialExpression.SAD, "I'm afraid not, only the big cheeses get to dine with the", "Duke.").also { stage++ }
            202 -> player(FacialExpression.NEUTRAL, "Well, maybe one day I'll be important enough to sit on", "the Duke's table.").also { stage ++ }
            203 -> npc(FacialExpression.NEUTRAL, "Maybe, but I won't be holding my breath.").also { stage++ }

            //Activate the Cook's Assistant Quest Complete Certificate
            204 -> end().also { player?.questRepository?.getQuest("Cook's Assistant")?.finish(player!!) }

            //Dialogue after Cook's Assistant Completion
            300 -> if(player.questRepository.getQuest("Lost Tribe").getStage(player) == 10) {
                player("Do you know what happened in the castle cellar?").also { stage = 600 }
            } else {
                    options("I am getting strong and mighty.", "I keep on dying.", "Can I use your range?").also { stage++ }
                   }
            301 -> when (buttonId) {
                1 -> player(FacialExpression.HAPPY, "I am getting strong and mighty. Grrr").also { stage = 310 }
                2 -> player(FacialExpression.SAD, "I keep on dying.").also { stage = 320 }
                3 -> player(FacialExpression.ASKING, "Can I use your range?").also { stage = 330 }
            }

            //I am getting strong and mighty
            310 -> npc(FacialExpression.HAPPY,"Glad to hear it!").also { stage = 1000 }

            //I keep on dying
            320 -> npc(FacialExpression.HAPPY, "Ah, well, at least you keep coming back to life too!").also { stage = 1000 }

            //Can I use your range?
            330 -> npc(FacialExpression.HAPPY, "Go ahead! It's a very good range; it's better than most" ,"other ranges.").also { stage++ }
            331 -> npc(FacialExpression.NEUTRAL, "It's called the Cook-o-Matic 100 and it uses a combination","of state-of-the-art temperature regulation and magic.").also { stage++ }
            332 -> player(FacialExpression.ASKING,"Will it mean my food will burn less often?").also { stage++ }
            333 -> npc(FacialExpression.NEUTRAL,"Well, that's what the salesman told us anyway...").also { stage++ }
            334 -> player(FacialExpression.THINKING, "Thanks?").also { stage = 1000 }
            //Conversation Endpoint
            1000 -> end()
        }
        return true
    }

    override fun newInstance(player: Player?): DialoguePlugin {
        return LumbridgeCookDialogue(player)
    }

    override fun getIds(): IntArray {
        return intArrayOf(278)
    }
}